File: implementations/dart/test/i57_test.dart
import 'package:test/test.dart';
import 'package:f57/b57.dart';

void main() {
  group('I57', () {
    test('encode/decode', () {
      final input = 'hello world'.codeUnits;
      final enc = i57Encode(input);
      final dec = i57Decode(enc);
      expect(dec, input);
    });

    test('hash and id', () {
      final h = i57Hash('x'.codeUnits, H57Length.len128);
      expect(h.isNotEmpty, true);

      final id = i57Id('x'.codeUnits, ID57Length.len128);
      expect(id.isNotEmpty, true);
    });

    test('entropy heuristics', () {
      final id = i57Random(22);
      expect(id.isNotEmpty, true);
    });

    group('i57ValidateIdentifier', () {
      test('bit-length branch: accepts a valid identifier, rejects corruption', () {
        final input = 'i57-validate-bitlength'.codeUnits;
        final id = i57Id(input, ID57Length.len128);

        expect(i57ValidateIdentifier(id, ID57Length.len128), true);
        // Default parameter also resolves to LEN_128.
        expect(i57ValidateIdentifier(id), true);

        // Truncated/corrupted string: too short for LEN_128's bound.
        final truncated = id.substring(0, 5);
        expect(i57ValidateIdentifier(truncated, ID57Length.len128), false);

        // Wrong bit length entirely.
        expect(i57ValidateIdentifier(id, ID57Length.len256), false);
      });

      test('fixed-width branch: accepts a valid identifier, rejects corruption', () {
        final input = 'i57-validate-fixedwidth'.codeUnits;
        final id = i57Id(input, ID57Length.fixed8);

        expect(id.length, 8);
        expect(i57ValidateIdentifier(id, ID57Length.fixed8), true);

        // Truncated/corrupted string: wrong length for FIXED_8.
        final truncated = id.substring(0, 4);
        expect(i57ValidateIdentifier(truncated, ID57Length.fixed8), false);

        // Corrupted character (invalid B57 alphabet char).
        final corrupted = '${id.substring(0, 7)}0';
        expect(i57ValidateIdentifier(corrupted, ID57Length.fixed8), false);
      });

      test('bit-length branch checks decoded byte length, not just char count', () {
        // 'A' * 22 is within LEN_128's [16,22] char bound and is valid,
        // canonical B57 - but it decodes to 22 zero bytes, not the 16
        // bytes LEN_128 requires. The byte-length check must reject it;
        // being in the char bound is not sufficient.
        expect(i57ValidateIdentifier('A' * 22, ID57Length.len128), false);
      });

      test('throws for undefined length_enum values', () {
        expect(
          () => i57ValidateIdentifier('x', -13),
          throwsA(isA<InvalidLengthEnumError>()),
        );
      });
    });
  });
}

File: implementations/ts/i57.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';
import { 
  i57Encode,
  i57Decode,
  i57Hash,
  i57Random,
  i57Id,
  i57IsValid,
  i57IsCanonical,
  i57ValidateIdentifier,
  i57ValidateEntropy
} from './i57.js';
import { R57Mode } from './r57.js';

test('i57Encode and i57Decode', () => {
  const input = new Uint8Array([1, 2, 3]);
  const enc = i57Encode(input);
  assert.equal(typeof enc, 'string');
  const dec = i57Decode(enc);
  assert.deepEqual(Array.from(dec), Array.from(input));
});

test('i57Hash', () => {
  const input = new Uint8Array([1, 2, 3]);
  const res = i57Hash(input, 47);
  assert.equal(typeof res, 'string');
});

test('i57Random', () => {
  const res = i57Random(R57Mode.R57_1_CSPRNG);
  assert.equal(typeof res, 'string');
});

test('i57Id', () => {
  const input = new Uint8Array([1, 2, 3]);
  const res = i57Id(input, -8); // FIXED_8: negative length_enum = fixed 8-char width
  assert.equal(typeof res, 'string');
  assert.equal(res.length, 8);
});

test('i57IsValid', () => {
  assert.equal(i57IsValid('222'), true);
  assert.equal(i57IsValid('O'), false);
  assert.equal(i57IsValid(''), false);
});

test('i57IsCanonical', () => {
  assert.equal(i57IsCanonical('222'), true);
  assert.equal(i57IsCanonical('O'), false);
  assert.equal(i57IsCanonical(''), false);
});

test('i57ValidateIdentifier', () => {
  const id = i57Random(R57Mode.R57_1_CSPRNG);
  assert.equal(i57ValidateIdentifier(id), true);
  assert.equal(i57ValidateIdentifier('short'), false);
  assert.equal(i57ValidateIdentifier('12345678901234567890 0'), false);
});

test('i57ValidateIdentifier dispatches on length_enum sign', () => {
  const input = new Uint8Array([1, 2, 3]);

  // Fixed width (negative): delegates to id57IsLength - exact width + alphabet only.
  const fixed = i57Id(input, -8);
  assert.equal(i57ValidateIdentifier(fixed, -8), true);
  assert.equal(i57ValidateIdentifier(fixed.slice(0, 7), -8), false);

  // Bit length (positive): bound + canonical + decoded byte length.
  const bitLen = i57Id(input, 32);
  assert.equal(i57ValidateIdentifier(bitLen, 32), true);
  assert.equal(i57ValidateIdentifier(bitLen.slice(0, 1), 32), false);
});

test('i57ValidateEntropy', () => {
  const id = i57Random(R57Mode.R57_1_CSPRNG);
  assert.equal(i57ValidateEntropy(id), true);
  assert.equal(i57ValidateEntropy('AAAAAAAAAAAAAAAAAAAAAA'), false);
  assert.equal(i57ValidateEntropy('ABABABABABABABABABABAB'), false);
});

